server/rpc_server.go

```diff
@@ -518,8 +527,25 @@ func (s *rpcServer) Start() error {
 			select {
 			// register self on interval
 			case <-t.C:
+				// if register check func defined, use it
+				if fn := s.opts.RegisterCheck; fn != nil {
+					if err = fn(s.opts.Context); err != nil {
+						log.Logf("Server %s-%s register error: %s", config.Name, config.Id, err)
+						s.RLock()
+						registered := s.registered
+						s.RUnlock()
+						if registered {
+							// deregister self in case of error
+							if err := s.Deregister(); err != nil {
```

Comment:
You've not deregistered and not set s.registered to false. This code block is also a bit unclean in a huge if statement. Maybe there should be a default check function that always returns nil so we don't actually have to test the function exists.

Reply:
i think that i fix this with new commit